### ISSUE TITLE
Repository-hdfs forbiddenApisJavaRestTest fails occasionally with permission denied error

### ISSUE THREAD

**Issue by masseyke** [elastic/elasticsearch#101524]
### CI Link

https://gradle-enterprise.elastic.co/s/yecause3qdmxi

### Repro line

./gradlew :plugins:repository-hdfs:forbiddenApisJavaRestTest

### Does it reproduce?

No

### Applicable branches

main

### Failure excerpt

It looks like every once in a while we get a `permission denied` error when trying to read `hdfs_hdfs.build.elastic.co.keytab`. It looks like `org.gradle.internal.hash.DefaultFileHasher` assumes that the cause of the  `FileNotFoundException` is that the file does not exist, but the actual error reported is that we don't have permission to read it.
```
Caused by: java.io.UncheckedIOException: Failed to create MD5 hash for file '/dev/shm/elastic+elasticsearch+main+intake+multijob+part1/test/fixtures/krb5kdc-fixture/testfixtures_shared/shared/hdfs/keytabs/hdfs_hdfs.build.elastic.co.keytab' as it does not exist.
	at org.gradle.internal.hash.DefaultFileHasher.hash(DefaultFileHasher.java:38)
	at org.gradle.api.internal.changedetection.state.CachingFileHasher.snapshot(CachingFileHasher.java:94)
	at org.gradle.api.internal.changedetection.state.CachingFileHasher.hash(CachingFileHasher.java:76)
	at org.gradle.api.internal.changedetection.state.SplitFileHasher.hash(SplitFileHasher.java:54)
	at org.gradle.internal.snapshot.impl.DirectorySnapshotter$PathVisitor.snapshotFile(DirectorySnapshotter.java:425)
	at org.gradle.internal.snapshot.impl.DirectorySnapshotter$PathVisitor.visitResolvedFile(DirectorySnapshotter.java:393)
	at org.gradle.internal.snapshot.impl.DirectorySnapshotter$PathVisitor.doVisitFile(DirectorySnapshotter.java:353)
	at org.gradle.internal.snapshot.impl.DirectorySnapshotterStatistics$CollectingFileVisitor.visitFile(DirectorySnapshotterStatistics.java:125)
	at org.gradle.internal.snapshot.impl.DirectorySnapshotterStatistics$CollectingFileVisitor.visitFile(DirectorySnapshotterStatistics.java:106)
	at java.base/java.nio.file.Files.walkFileTree(Files.java:2811)
	at org.gradle.internal.snapshot.impl.DirectorySnapshotter.snapshot(DirectorySnapshotter.java:119)
	at org.gradle.internal.vfs.impl.DefaultFileSystemAccess.lambda$snapshot$8(DefaultFileSystemAccess.java:155)
	at org.gradle.internal.vfs.impl.AbstractVirtualFileSystem.store(AbstractVirtualFileSystem.java:84)
	at org.gradle.internal.vfs.impl.DefaultFileSystemAccess.snapshot(DefaultFileSystemAccess.java:145)
	at org.gradle.internal.vfs.impl.DefaultFileSystemAccess.lambda$read$0(DefaultFileSystemAccess.java:82)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at org.gradle.internal.vfs.impl.DefaultFileSystemAccess.lambda$readSnapshotFromLocation$9(DefaultFileSystemAccess.java:187)
	at org.gradle.internal.vfs.impl.DefaultFileSystemAccess$StripedProducerGuard.guardByKey(DefaultFileSystemAccess.java:221)
	at org.gradle.internal.vfs.impl.DefaultFileSystemAccess.lambda$readSnapshotFromLocation$10(DefaultFileSystemAccess.java:184)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at org.gradle.internal.vfs.impl.DefaultFileSystemAccess.readSnapshotFromLocation(DefaultFileSystemAccess.java:184)
	at org.gradle.internal.vfs.impl.DefaultFileSystemAccess.readSnapshotFromLocation(DefaultFileSystemAccess.java:169)
	at org.gradle.internal.vfs.impl.DefaultFileSystemAccess.read(DefaultFileSystemAccess.java:82)
	at org.gradle.internal.fingerprint.impl.DefaultFileCollectionSnapshotter$SnapshottingVisitor.visitCollection(DefaultFileCollectionSnapshotter.java:70)
	at org.gradle.api.internal.artifacts.ivyservice.ResolvedFileCollectionVisitor.endVisitCollection(ResolvedFileCollectionVisitor.java:57)
	at org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ArtifactVisitorToResolvedFileVisitorAdapter.endVisitCollection(ArtifactVisitorToResolvedFileVisitorAdapter.java:56)
	at org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ArtifactBackedResolvedVariant$SingleArtifactSet.visit(ArtifactBackedResolvedVariant.java:151)
	at org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ParallelResolveArtifactSet$VisitingSet$StartVisitAction.visitResults(ParallelResolveArtifactSet.java:100)
	at org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ParallelResolveArtifactSet$VisitingSet.visit(ParallelResolveArtifactSet.java:69)
	at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration.visitArtifacts(DefaultLenientConfiguration.java:302)
	at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration.access$600(DefaultLenientConfiguration.java:75)
	at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration$3.run(DefaultLenientConfiguration.java:249)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:73)
	at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration.visitArtifactsWithBuildOperation(DefaultLenientConfiguration.java:246)
	at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration.access$300(DefaultLenientConfiguration.java:75)
	at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration$2.lambda$visitArtifacts$0(DefaultLenientConfiguration.java:149)
	at org.gradle.internal.work.DefaultWorkerLeaseService.runAsUnmanagedWorkerThread(DefaultWorkerLeaseService.java:140)
	at org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration$2.visitArtifacts(DefaultLenientConfiguration.java:149)
	at org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.SelectedArtifactSet.visitFiles(SelectedArtifactSet.java:34)
	at org.gradle.api.internal.artifacts.configurations.ResolutionBackedFileCollection.visitContents(ResolutionBackedFileCollection.java:72)
	at org.gradle.api.internal.file.AbstractFileCollection.visitStructure(AbstractFileCollection.java:366)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.visitContents(DefaultConfiguration.java:560)
	at org.gradle.api.internal.file.AbstractFileCollection.visitStructure(AbstractFileCollection.java:366)
	at org.gradle.api.internal.file.CompositeFileCollection.lambda$visitContents$0(CompositeFileCollection.java:133)
	at org.gradle.api.internal.file.UnionFileCollection.visitChildren(UnionFileCollection.java:81)
	at org.gradle.api.internal.file.CompositeFileCollection.visitContents(CompositeFileCollection.java:133)
	at org.gradle.api.internal.file.AbstractFileCollection.visitStructure(AbstractFileCollection.java:366)
	at org.gradle.api.internal.file.CompositeFileCollection.lambda$visitContents$0(CompositeFileCollection.java:133)
	at org.gradle.api.internal.file.UnionFileCollection.visitChildren(UnionFileCollection.java:81)
	at org.gradle.api.internal.file.CompositeFileCollection.visitContents(CompositeFileCollection.java:133)
	at org.gradle.api.internal.file.AbstractFileCollection.visitStructure(AbstractFileCollection.java:366)
	at org.gradle.api.internal.file.CompositeFileCollection.lambda$visitContents$0(CompositeFileCollection.java:133)
	at org.gradle.api.internal.file.UnionFileCollection.visitChildren(UnionFileCollection.java:81)
	at org.gradle.api.internal.file.CompositeFileCollection.visitContents(CompositeFileCollection.java:133)
	at org.gradle.api.internal.file.AbstractFileCollection.visitStructure(AbstractFileCollection.java:366)
	at org.gradle.api.internal.file.CompositeFileCollection.lambda$visitContents$0(CompositeFileCollection.java:133)
	at org.gradle.api.internal.file.UnionFileCollection.visitChildren(UnionFileCollection.java:81)
	at org.gradle.api.internal.file.CompositeFileCollection.visitContents(CompositeFileCollection.java:133)
	at org.gradle.api.internal.file.AbstractFileCollection.visitStructure(AbstractFileCollection.java:366)
	at org.gradle.api.internal.file.CompositeFileCollection.lambda$visitContents$0(CompositeFileCollection.java:133)
	at org.gradle.api.internal.file.UnionFileCollection.visitChildren(UnionFileCollection.java:81)
	at org.gradle.api.internal.file.CompositeFileCollection.visitContents(CompositeFileCollection.java:133)
	at org.gradle.api.internal.file.AbstractFileCollection.visitStructure(AbstractFileCollection.java:366)
	at org.gradle.api.internal.file.CompositeFileCollection.lambda$visitContents$0(CompositeFileCollection.java:133)
	at org.gradle.api.internal.file.collections.UnpackingVisitor.add(UnpackingVisitor.java:67)
	at org.gradle.api.internal.file.collections.UnpackingVisitor.add(UnpackingVisitor.java:92)
	at org.gradle.api.internal.file.DefaultFileCollectionFactory$ResolvingFileCollection.visitChildren(DefaultFileCollectionFactory.java:285)
	at org.gradle.api.internal.file.CompositeFileCollection.visitContents(CompositeFileCollection.java:133)
	at org.gradle.api.internal.file.AbstractFileCollection.visitStructure(AbstractFileCollection.java:366)
	at org.gradle.api.internal.file.CompositeFileCollection.lambda$visitContents$0(CompositeFileCollection.java:133)
	at org.gradle.api.internal.tasks.PropertyFileCollection.visitChildren(PropertyFileCollection.java:48)
	at org.gradle.api.internal.file.CompositeFileCollection.visitContents(CompositeFileCollection.java:133)
	at org.gradle.api.internal.file.AbstractFileCollection.visitStructure(AbstractFileCollection.java:366)
	at org.gradle.internal.fingerprint.impl.DefaultFileCollectionSnapshotter.snapshot(DefaultFileCollectionSnapshotter.java:47)
	at org.gradle.internal.execution.impl.DefaultInputFingerprinter$InputCollectingVisitor.visitInputFileProperty(DefaultInputFingerprinter.java:133)
	at org.gradle.api.internal.tasks.execution.TaskExecution.visitRegularInputs(TaskExecution.java:328)
	... 74 more
Caused by: java.io.FileNotFoundException: /dev/shm/elastic+elasticsearch+main+intake+multijob+part1/test/fixtures/krb5kdc-fixture/testfixtures_shared/shared/hdfs/keytabs/hdfs_hdfs.build.elastic.co.keytab (Permission denied)
	at java.base/java.io.FileInputStream.open0(Native Method)
	at java.base/java.io.FileInputStream.open(FileInputStream.java:216)
	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:157)
	at org.gradle.internal.hash.DefaultFileHasher.hash(DefaultFileHasher.java:36)
	... 156 more
```

**Comment by ywangd** [comment]
IIUC, HDFS is still owned by data management. I will change the team label accordingly. Also the failure is for forbiddenApi. The error stack trace is entirely within Gradle. So I'll adjust this to "low-risk".

**Comment by jbaiera** [comment]
I've spent quite a bit of time today looking into this failure and noticed a couple of weird things about it. When we execute the kerberos test fixture, it stands up a container running the KDC and then provisions the principals. The postProcessFixture task then executes which should copy the keytabs and krb5.conf files to a local directory for usage in the tests.

In all my local executions the post process task executes very shortly after the container is stood up, but that does not seem to be the case on the runs that failed in CI:

Example from https://gradle-enterprise.elastic.co/s/2nrivkghqhvlo/console-log?page=9#L8358-L8376 shows that the forbidden API's task fails before the krb5 fixture is even fully stood up. 
https://gradle-enterprise.elastic.co/s/2nrivkghqhvlo/console-log?page=9#L8410 from that same build shows that the post processing task has only run some 30 odd lines later. 

Another example shows a failure at https://buildkite.com/elastic/elasticsearch-periodic/builds/1036#018c5c66-ebd9-40ee-bfed-4d466d70de67/6-5157. In that same build, we can see that the fixture is running beforehand https://buildkite.com/elastic/elasticsearch-periodic/builds/1036#018c5c66-ebd9-40ee-bfed-4d466d70de67/6-5094, but no post process operation occurs before the dependent task fails.

This leads me to a couple questions that are still outstanding:

1) Why does the forbidden API task need the keytab and conf files at all? I don't believe they contribute anything to the task.
2) Where is the forbidden API task getting those dependencies from? Exploring the build file for hdfs plugin shows that those krb5 files are pulled into their own separate configurations. I notice there are some lines that hint that the keytabs might find their way into some runtime classpaths, but I do not see where that could possibly be occurring from the build file... Maybe I'm just not looking in the right places?
3) If the keytab configuration is finding its way into the forbidden API task's classpath, then why is Gradle not resolving it? Those files are marked as artifacts in the krb5 project and thus Gradle should know that the post processing task creates them, but it does not seem to notice that in these cases, which makes me wonder if a plain path is being added somewhere. I've searched through the build script on the hdfs plugin project but could not find anything that was amiss, though once again, I may be missing an obvious Gradle antipattern.
4) The initial error mentions that reading the file causes a permission error, but I'm not entirely sure if that's the case. On my local build, I have found that those keytab files are cleaned up at the end of every build, so they would not actually be present unless the post process operation is done. 

All in all, I don't think this is actually a hadoop problem, but more of a tooling problem for the build. I'd like to attach the delivery team to the issue to see what their read on the problem might be.

cc @elastic/es-delivery
